Another batch of changes

tomers opened the pull request:
Each commit does something else, so please read each one separately

maresb commented:
Thanks @tomers!

I'm not quite following what is `VERTICAL_PREVIEW_MARGIN_PX`. Could you explain it?

Is the idea that for the preview we add a fake margin for the nonprintable area so that the preview resembles the physical tape after printing?

maresb commented:
That's really nice with the adjustable margin in the preview. But it's a bit weird that the margin code for the preview is different from the margin code for the printer.

Looking more closely I see that the printer code prints no margin at the beginning and a double-margin at the end. I think this probably has to do with the gap between the printer head and the cutter. I'm guessing that the gap is approximately `DEFAULT_MARGIN_PX` pixels long. Then when we print `2 * DEFAULT_MARGIN_PX`, where the first `DEFAULT_MARGIN_PX` is the true right-margin of the current label, and the second `DEFAULT_MARGIN_PX` is the left-margin of the next label. Do you agree with this assessment?

It might be nice to start distinguishing some of these concepts in the code. Like have a constant for `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX` = 13, and have separate variables in the code, something like `current_right_margin` and `future_left_margin` (or whatever makes sense), just to make the code somewhat consistent with reality. 

In principle we could enable a margin-free mode where you print the first `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX` pixels, tell the user to cut, and print the remaining pixels plus future margin, and tell the user to cut again.

If the user wants a label with `margin` greater than `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX`, then I think the correct way to do this is:
* Print a left margin of `additional_left_margin = margin - GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX`, since we already have `GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX` due to the gap from the previous label.
* Print the main payload
* Print a right margin of `margin`
* Continue printing a margin of `future_left_margin = GAP_BETWEEN_PRINT_HEAD_AND_CUTTER_PX`.
(I think the current implementation is buggy since it'll produce too short of a left-margin and too long of a right-margin, but I'm not able to test at the moment.)

Am I making sense?

tomers commented:
@marseb, thanks for the insights. I've contrinued to modify the code, and I am going to introduce some further changes soon. Thanks for the insight of explaining why we print the margins twice at the end. I was puzzled by that.
I will add some of your text as comments in the code, and I also consider taking the variable and constant names you suggest. Let's keep this PR open for a while, until I'm done with that.

maresb commented:
Awesome, thanks so much for looking into that!!!

I've converted the PR to draft for now. When you're finished just hit "Ready for review" and I'll give it another look and merge if I approve.

tomers commented:
Exactly. I still need to work on that part, but the idea is that preview
should show both horizontal and vertical margins.

On Sat, Feb 3, 2024, 12:38 Ben Mares ***@***.***> wrote:

> Thanks @tomers <https://github.com/tomers>!
>
> I'm not quite following what is VERTICAL_PREVIEW_MARGIN_PX. Could you
> explain it?
>
> Is the idea that for the preview we add a fake margin for the nonprintable
> area so that the preview resembles the physical tape after printing?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/computerlyrik/dymoprint/pull/113#issuecomment-1925271482>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUL46ACLSJ36PFVNAFMEDYRYHRDAVCNFSM6AAAAABCXDDOQ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMRVGI3TCNBYGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maresb commented:
Okay, great!

Keep in mind that there are a few different levels of abstraction coming together:
* the bitmap itself that's to be printed
* the way the bitmap is printed onto a piece of tape

I think there are valid use cases to have preview functionality for both. Unfortunately I don't have any clever ideas for an implementation.

Note that this is starting to connect up with @tomek-szczesny's work on a lookup table for various printer and label width combinations in #91.

It's really great how things are coming together thanks to your work @tomers! Back when #91 was opened, I don't think the code was healthy enough to carry through that implementation. But now the code is becoming much easier to understand, opening up lots of exciting possibilities!

tomek-szczesny commented:
Certainly I'm glad someone's cleaning up my amateur code. :) 
Just wanted to add, I regret that printer must be detected before rendering, in order to determine the actual printable area. I see no other way around it, barren the user-provided printer model as a command line parameter.

maresb commented:
> my amateur code

To be clear the obstacle wasn't your code, it was instead the underlying code which was very idiosyncratic with not-so-great separation of concerns and idiosyncratic variable names. Those sorts of things build up and create a tremendous amount of friction for making changes.

tomers commented:
Thanks, @maresb and @tomek-szczesny!
I'm working on discarding margin handling in the printing driver. This adds mental load on understanding the code, and the principle of having to end the print job with an offset for the next print job, and will be a burden in the future, when we'll have multiple drivers (which would need to duplicate this task), and also in the case where we may want to implement some more complicated scenarios, such as printing batch of labels, with zero margin (i.e. print brank without ending with offset, cut the label, then iteratively printing labels without margins, or at least with SW defined margins, and instructing user to cut).

@maresb, regarding your comment about idiosyncratic code, that's indeed true, and the was my initial effort, to clean these things off the table, before approaching any other new features.

maresb commented on src/dymoprint/lib/utils.py:
```diff
@@ -33,12 +33,16 @@ def draw_image(bitmap):
     return math.ceil(mm * 10) / 10
 
 
+def mm_to_px(mm) -> float:
```
```suggestion
def mm_to_px(mm) -> int:
```

tomers replied:
I prefer to have pixel as float still. It indeed makes no physical sense, but helps with calculations along the way. I prefer to convert to int at the very end.

tomers replied:
Added one more commit. Please review.

tomers commented:
Hey @maresb and @tomek-szczesny,

Can you please review this PR?

Thanks,
Tomer

maresb commented on src/dymoprint/lib/utils.py:
```diff
@@ -33,12 +33,16 @@ def draw_image(bitmap):
 
 
+def mm_to_px(mm) -> float:
+    return math.ceil(mm * PIXELS_PER_MM)
```
If you want it to return a float:

```suggestion
    return mm * PIXELS_PER_MM
```

maresb reviewed (changes requested):
I will finish a PR soon with some more in-depth suggestions

maresb commented:
Oh, looks like you're actively working on this. I'll hold off on further review for the moment.

maresb commented:
Are you okay with the conflicts there? Looks like maybe you accidentally rebased on a stale branch?

tomers commented:
It is ready for review now. Sorry, but I use `git rebase` a lot, and I keep forgeting to rebase on `origin/master`.

tomers commented:
Any news?

maresb commented:
Sorry, I'm in a time crunch at the moment. It might have to wait another few days.

tomers commented:
Hey there,

Due to irresponsivenes, I'm abandoning this project, and will fork it in a new organization, managed by me. Whoever wants to join me, is most welcome to contact me via LinkedIn.

maresb commented:
Dang, that's pretty drastic. As you wish.

tomers commented:
I hope I did not sound too drastic. That was not the intention. I mean that
I will cease development on that repository in the future.
I have no hard feelings at all! It is simply that I can't wait 3 weeks for
a PR to be merged... It's hard to work like that.
Note that my action is aligned to what we've discussed earlier this month
on LinkedIn. We agreed that we should open a separate, new, Github
organization for this purpose.


On Wed, Feb 28, 2024 at 10:04 PM Ben Mares ***@***.***> wrote:

> Dang, that's pretty drastic. As you wish.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/computerlyrik/dymoprint/pull/113#issuecomment-1969774315>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUL4Y2Q352EJWMSIT3ZI3YV6EUNAVCNFSM6AAAAABCXDDOQ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRZG43TIMZRGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maresb commented:
Ya, understandable. I'm going through a very stressful period at the moment, and unfortunately I haven't had any time in the past few weeks to commit to open-source.

I would have expected such an announcement to be a bit more coordinated and communal, not so unilateral. But you seem to have a lot of energy, so I think it's good that you to take charge.

tomek-szczesny commented:
This repository is already abandoned by its actual owner so it does make sense to fork it into a new entity that will be easier to pass control to. I remember that @maresb had problems with privileges here already.

I think a link to the new repo / organization should be added on top of README here, and this repository archived if possible.

tomers commented:
> I would have expected such an announcement to be a bit more coordinated
and communal, not so unilateral
I totally agree with that. I am sorry for being hasty with my response.
Ben and Tomek, maybe we should find some time for a short discussion (this
medium is bad for that) about how we're going to push this project forward.
Alternatively, if you approve, I can open a new Github organization and
fork this project (I will also update README accordingly).

On Thu, Feb 29, 2024 at 6:43 AM Tomek Szczęsny ***@***.***>
wrote:

> This repository is already abandoned by its actual owner so it does make
> sense to fork it into a new entity that will be easier to pass control to.
> I remember that @maresb <https://github.com/maresb> had problems with
> privileges here already.
>
> I think a link to the new repo / organization should be added on top of
> README here, and this repository archived if possible.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/computerlyrik/dymoprint/pull/113#issuecomment-1970385605>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUL47UKIZC3HZ7OP7FPELYV2YVZAVCNFSM6AAAAABCXDDOQ6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZQGM4DKNRQGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tomek-szczesny commented:
I don't feel there is anything left to discuss, a new organization and fork of this repository is reasonable and I think everyone has already expressed approval. Just let us know so we can join it. :)

maresb commented:
I wanted to send an email to recent contributors, and also @computerlyrik, letting them know about the plan, and also to solicit a new name for the fork. There are also some logistical issues like how to communicate the deprecation of this project. I think a quick call may be useful to sort out these details. @tomek-szczesny, it'd be nice to have you there too, but @tomers and I could probably make do with just the two of us.

tomers commented:
@tomek-szczesny, @maresb, let's continue this discussion in issue #114 .

maresb commented:
Some preliminary review to fix some breakage that I'm seeing: https://github.com/tomers/dymoprint/pull/6

Here's the output of `dymoprint --preview asdf` on `v2.4.0`:

![image](https://github.com/computerlyrik/dymoprint/assets/15216687/0ed87e3c-c2fa-4d53-bfc2-5138bbab3409)

Here's the output on https://github.com/tomers/dymoprint/pull/6

![image](https://github.com/computerlyrik/dymoprint/assets/15216687/80bb4981-9af8-4e86-984d-95085d523169)

I can't even take a screenshot of the full output. I understand that we're enlarging the preview to simulate the printable area, but I'm wondering what all this extra black space is about.

maresb commented on src/dymoprint/gui/gui.py:
```diff
@@ -20,31 +21,28 @@
-    DEFAULT_TAPE_SIZE_MM_INDEX = 1
-
-    label_bitmap: Optional[Image.Image]
+    print_label_bitmap: Optional[Image.Image]
```
I read `print` as a verb, and so this reads like a method name rather than an attribute. Maybe rename to something like `label_bitmap_to_print`?

maresb commented on src/dymoprint/gui/q_dymo_labels_list.py:
```diff
@@ -79,48 +105,98 @@ def dropEvent(self, e) -> None:
+        )
+        try:
+            bitmap = render_engine.render(self.render_context)
+        except BaseException as err:  # noqa: BLE001
```
I'd strongly recommend against introducing this and the next blind except a few lines down. They make things extremely difficult to maintain down the line.

Even if it's a generic catch-all thing, at least make an effort to catch all the common errors and leave `BaseException` as a failsafe rather than a primary measure.

maresb commented on src/dymoprint/lib/dymo_labeler.py:
```diff
@@ -219,40 +213,41 @@ def _get_status(self):
         tape_size_mm: int = DEFAULT_TAPE_SIZE_MM,
     ):
-        self.margin_px = margin_px
+        assert tape_size_mm in self.SUPPORTED_TAPE_SIZES_MM
```
Unless there's something else forcing `tape_size_mm` to be a supported tape size, this should be a `ValueError` and not an `AssertionError`. Note that Python strips assert statements when run in optimized mode, so assert statements shouldn't be used for any sort of error handling.

(Note that given our installation instructions probably 100% of our users will not be running in optimized mode, so my statement is more on principle than in practice; semantically `ValueError` is more accurate than `AssertionError`.)

maresb commented on src/dymoprint/lib/unicode_blocks.py:
```diff
@@ -5,25 +5,33 @@
 
+BLACK_PIXEL = (255, 255, 255, 255)
+WHITE_PIXEL = (0, 0, 0, 255)
+TRANSPARENT_PIXEL = (0, 0, 0, 0)
```
What are transparent pixels for? Probably some explanation of this and `LS` should be added to the code as comments.

maresb commented on src/dymoprint/lib/render_engines/print_preview.py:
```diff
@@ -0,0 +1,205 @@
+                pixdata[x, y] = pixel_map[pixdata[x, y]]
+        return bitmap, meta
+
+    def _show_margins(self, label_bitmap, preview_bitmap, meta, context):
```
The result looks a bit off:
![image](https://github.com/computerlyrik/dymoprint/assets/15216687/83005fcc-ac2d-4b06-ac60-49ad99c421ec)

The text is white and difficult to read.
The vertical margin appears to be displayed horizontally under the vertical margin.

tomers replied:
I fixed the coloring. I've hard-coded the colors for now. The reason I've use the original colors is because I worked in dark mode, and it looks different there. I tried to choose colors which would look good on both modes, but this is still a WIP.

> The vertical margin appears to be displayed horizontally under the vertical margin
I am not sure I understand this sentence. Can you elaborate?

maresb replied:
Ah, now I see. It is really difficult to find something that works in both.

I don't know much about PyQt5, but I wonder if there's some way to set the colors conditionally based on light vs dark.

tomers replied:
I added a commit which addresses this issue

maresb reviewed (commented):
To avoid merge conflicts, please first see my suggested changes under https://github.com/tomers/dymoprint/pull/6.

tomers commented:
Hello @maresb , thank you for reviewing my code! :-)
Apologies for the delay in addressing the issues raised. Your feedback guided me towards better solutions, and I appreciate your patience throughout the process.

maresb commented:
More image mode headaches :dizzy_face: 

```python
$ dymoprint --browser asdf --verbose
[DEBUG] Read config file: /home/mares/.config/dymoprint.ini
[DEBUG] Demo mode: showing label..
[ERROR] not supported for this image mode
Traceback (most recent call last):
  File "/home/mares/repos/dymoprint/src/dymoprint/lib/utils.py", line 49, in system_run
    yield
  File "/home/mares/repos/dymoprint/src/dymoprint/cli/cli.py", line 343, in main
    run()
  File "/home/mares/repos/dymoprint/src/dymoprint/cli/cli.py", line 325, in run
    ImageOps.invert(bitmap).save(fp)
  File "/home/mares/micromamba/envs/dymoprint-dev/lib/python3.8/site-packages/PIL/ImageOps.py", line 563, in invert
    return image.point(lut) if image.mode == "1" else _lut(image, lut)
  File "/home/mares/micromamba/envs/dymoprint-dev/lib/python3.8/site-packages/PIL/ImageOps.py", line 60, in _lut
    raise OSError(msg)
OSError: not supported for this image mode
```

maresb commented on src/dymoprint/gui/q_dymo_labels_list.py:
```diff
@@ -73,61 +103,114 @@ def dropEvent(self, e) -> None:
+        )
+        try:
+            bitmap = render_engine.render(self.render_context)
+        except BaseException as err:  # noqa: BLE001
```
Can we get by here with `RenderEngineException` or do we need `BaseException` to cover all the cases?

maresb commented on src/dymoprint/gui/q_dymo_labels_list.py:
```diff
@@ -73,61 +103,114 @@ def dropEvent(self, e) -> None:
         )
         try:
-            label_bitmap = render_engine.render(self.render_context)
+            bitmap, _ = render_engine.render(self.render_context)
```
Can we get on the next line with `RenderEngineException` or do we need `BaseException` to cover all the cases?

maresb commented:
The issue with the huge preview margins described in https://github.com/computerlyrik/dymoprint/pull/113#issuecomment-1975153786 is resolved. :rocket:

maresb commented on src/dymoprint/lib/render_engines/barcode_with_text.py:
```diff
@@ -1,13 +1,16 @@
-from dymoprint.lib.render_engines.render_engine import RenderEngine
+from dymoprint.lib.render_engines.render_engine import (
+    RenderEngine,
+    RenderEngineException,
```
This should fit on a single line. (Just remove the last trailing comma and reformat.)

tomers replied:
Ruff does not approve of your suggestion, and pre-commit fails.
Also, when I save in VS Code, the IDE auto formats the line for me.

maresb replied:
Indeed. Somehow the line looked deceivingly short to me.

maresb commented on src/dymoprint/lib/render_engines/margins.py:
```diff
@@ -0,0 +1,115 @@
+        super().__init__(msg)
+
+
+class MarginsMode(Enum):
+    PRINT = 1
+    PREVIEW = 2
```
I'd recommend replacing the enum with a string literal for simplicity.

tomers replied:
I prefer not to. I think it is more obvious to have a single place where the possible values are.
I think this way it is better than having to add `Literal` type hint to each argument in the function call chain, where in each such type hint we have to specify all accepted values. This is what `Enum`s are for, no?

maresb replied:
Yes, that's what `Enum`s are for, but I disagree that it's more obvious to have the possible values in a single place.

Probably the most important metric of code quality to me is, "How much do I have to poke around in the code before I understand what's going on?" When using an `Enum` as opposed to `Literal` it means that the given block of code no longer stands on its own. Whenever I want to understand or use it (e.g. from a REPL) I must look into another file to read the definition and then either import the `Enum` or make use of the meaningless correspondence between names and values.

The obvious counterpoint is `Enum`s are good because they're DRY, and DRY is good. But why is DRY good? It usually reduces code length, but I prioritize clarity over length. It's also good to prevent different parts of code from going out-of-sync. But with `Literal` you have the type checker for that. And the type checker will do a much better job with `Literal` than `Enum`. If you add a new member to your `Enum` you should update all functions that use it. With `Literal` the type checker should automatically flag all the functions that need to be updated. It's also more powerful because for example it's easy to define functions which only handle subsets of the keys.

`Enum`s aren't terrible, but unless you have a large number of elements or if there's a meaningful correspondence between names and values, I just don't see any benefit. My feeling is that they're mostly obsolete in modern Python. Ultimately it's a subjective style choice, so feel free to pass if you're unconvinced.

tomers replied:
I've commited a change to use literal type hint instead of enum. I feel less confident relying on the liner for detecting issues such as we had earlier during review, when in some places I've forgot to update the type literal. I personally favor DRY code (I really struggle not to fix wet code whenever I encounter it), which I believe improves code's maintainability. I think that prioritizing maintanability over readability (simplicity of the code) is the way to go. However, in this project, I do not want to enforce my personal preference, and I have no problem compromising to get to a consensus, which I believe would increase productivity and improve overall developer experience of all of us :-)

maresb commented on src/dymoprint/lib/render_engines/margins.py:
```diff
@@ -0,0 +1,115 @@
+        self,
+        render_engine: RenderEngine,
+        mode: MarginsMode,
+        justify: str = "center",
```
```suggestion
        justify: Literal["left", "center", "right"] = "center",
```

maresb commented on src/dymoprint/lib/render_engines/margins.py:
```diff
@@ -0,0 +1,115 @@
+        # There is a gap between the printer head and the cutter (for the sake of this
+        # example, let us say it is DX pixels wide).
+        # We assume the printing starts when the print head is in offset DX from the
+        # label's edge (just under the cutter).
```
```suggestion
        # label's edge (the label's edge begins just under the cutter).
```

maresb replied:
What do you think of this?

maresb commented on src/dymoprint/lib/render_engines/margins.py:
```diff
@@ -0,0 +1,115 @@
+        # After we print the payload, we need to offset the label DX pixels, in order
+        # to move the edge of the printed payload past the cutter, otherwise the cutter
+        # will cut inside the printed payload.
+        # Afterwards, we need to offset another DX pixels, so that the cut will have
+        # some margin from the payload edge. The reason we move DX pixels this time, is
+        # in order to have symmetry with the initial margin between label edge and start
+        # of printed payload.
```
This offset business is so confusing. I'm trying to make it a bit more clear:

```suggestion
        # Subsequently, in order to achieve symmetry so that the final margin matches
        # the initial margin, we add another offset of DX pixels, so that the cut
        # will have that same margin from both sides of the payload edge. In summary,
        # due to the gap between the printer head and the cutter of DX pixels,
        # printing an initial margin of 0 pixels and a final margin of 2 * DX pixels
        # leaves an effective margin of DX pixels on both sides.
```

maresb replied:
And this?

tomers replied:
Perfect. Thanks.

maresb commented on src/dymoprint/lib/render_engines/print_payload.py:
```diff
@@ -0,0 +1,32 @@
+    def __init__(
+        self,
+        render_engine: RenderEngine,
+        justify: str = "center",
```
We should probably do a search through the codebase for "left" and "center" and make sure everything's typed correctly.
```suggestion
        justify: Literal["left", "center", "right"] = "center",
```

maresb replied:
There's still one missing in `dymoprint.gui.q_dymo_labels_list` line 116.

tomers replied:
Fixed.
That's exactly what I dislike about working wet, without Enums... You have to update 100 places.
I've searched in the code, and I think I've updated in all places.

maresb replied:
Seems to me like a feature. Sometimes a change does require an update in 100 places, and this makes it explicit. :slightly_smiling_face:

maresb commented on src/dymoprint/lib/render_engines/print_preview.py:
```diff
@@ -0,0 +1,201 @@
+    def __init__(
+        self,
+        render_engine: RenderEngine,
+        justify: str = "center",
```
```suggestion
        justify: Literal["left", "center", "right"] = "center",
```

maresb commented on src/dymoprint/lib/unicode_blocks.py:
```diff
@@ -5,25 +5,35 @@
+    (BLACK_PIXEL, WHITE_PIXEL): UH,
+    (WHITE_PIXEL, WHITE_PIXEL): NB,
+    (BLACK_PIXEL, TRANSPARENT_PIXEL): UH,
+    (TRANSPARENT_PIXEL, TRANSPARENT_PIXEL): NB,
```
I believe only the second pixel can be transparent, right?
```suggestion
    (WHITE_PIXEL, TRANSPARENT_PIXEL): NB,
```

We should probably be testing this with an odd-length payload that actually touches the edge.

tomers replied:
`dymoprint --preview-inverted "single line"` should cover it (in current setup it results in an off image)

maresb commented on src/dymoprint/lib/unicode_blocks.py:
```diff
@@ -5,25 +5,35 @@
+    (WHITE_PIXEL, BLACK_PIXEL): UH,
+    (BLACK_PIXEL, WHITE_PIXEL): LH,
+    (WHITE_PIXEL, WHITE_PIXEL): FB,
+    (BLACK_PIXEL, TRANSPARENT_PIXEL): LS,
```
```suggestion
    (BLACK_PIXEL, TRANSPARENT_PIXEL): NB,
    (WHITE_PIXEL, TRANSPARENT_PIXEL): UH
```

maresb commented on src/dymoprint/lib/unicode_blocks.py:
```diff
@@ -5,25 +5,35 @@
 assert LH == "\N{LOWER HALF BLOCK}"
 assert FB == "\N{FULL BLOCK}"
 assert NB == "\N{NO-BREAK SPACE}"
+assert LS == "\N{LIGHT SHADE}"
```
Unfortunately it doesn't seem like there are unicode block characters that shade the upper and lower half independently.

It seems to me like we're probably better off leaving the transparent pixels unshaded.

Unless I'm confused!

```suggestion
```

tomers replied:
I just added the shade for debugging purposes (to see where it is used), and probably forgot to remove it.

tomers commented:
Done with the fixes, @maresb

maresb commented:
`--browser` now works again. There are still a few things unresolved from the last review.

It's really difficult to review such a large PR after a rebase, so could you please refrain from rebasing until after I've approved?

maresb reviewed (approved):
Whew, that was a whole lot of work, but I think we finally got it. Thanks @tomers for persevering. These are some really huge improvements.

Go ahead and rebase if you'd like, and then let's get this merged.

tomers commented:
Thanks Ben,

The branch is ready to be merged from my side. I am not going to change it
anymore. You can merge according to the plan that we discussed privately.

On Sun, Mar 24, 2024 at 8:13 PM Ben Mares ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> Whew, that was a whole lot of work, but I think we finally got it. Thanks
> @tomers <https://github.com/tomers> for persevering. These are some
> really huge improvements.
>
> Go ahead and rebase if you'd like, and then let's get this merged.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/computerlyrik/dymoprint/pull/113#pullrequestreview-1956616353>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAUL43WPD4PD2HOEJXUNDLYZ4JVBAVCNFSM6AAAAABCXDDOQ6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNJWGYYTMMZVGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

maresb commented:
Thanks @tomers!

For anyone reading, to avoid giving the impression of being secretive, the plan is to not release this as DymoPrint, but instead release it as Labelle v1.1.0 (see #114). And Labelle v1.0.0 will be a rebranded DymoPrint v2.4.0.
